src/emitters/spot.cpp

```diff
@@ -0,0 +1,179 @@
+#include <mitsuba/core/properties.h>
+#include <mitsuba/core/warp.h>
+#include <mitsuba/render/emitter.h>
+#include <mitsuba/render/medium.h>
+#include <mitsuba/render/texture.h>
+
+NAMESPACE_BEGIN(mitsuba)
+
+/**!
+
+.. _emitter-spot:
+
+Spot light source (:monosp:`spot`)
+------------------------------------
+
+.. pluginparameters::
+
+ * - intensity
+   - |spectrum|
+   - Specifies the maximum radiant intensity at the center in units of power per unit steradian. (Default: 1).
+ * - cutoff_angle
+   - |float|
+   - Cutoff angle, beyond which the spot light is completely black (Default: 20 degrees)
+ * - beam_width
+   - |float|
+   - Subtended angle of the central beam portion (Default: :math:`cutoff_angle \times 3/4`)
+ * - texture
+   - |texture|
+   - An optional texture to be projected along the spot light
+ * - to_world
+   - |transform|
+   - Specifies an optional emitter-to-world transformation.  (Default: none, i.e. emitter space = world space)
+
+This plugin provides a spot light with a linear falloff. In its local coordinate system, the spot light is
+positioned at the origin and points along the positive Z direction. It can be conveniently reoriented
+using the lookat tag, e.g.:
+
+.. code-block:: xml
+    :name: spot-light
+
+    <emitter type="spot">
+        <transform name="to_world">
+            <!-- Orient the light so that points from (1, 1, 1) towards (1, 2, 1) -->
+            <lookat origin="1, 1, 1" target="1, 2, 1"/>
+        </transform>
+    </emitter>
+
+The intensity linearly ramps up from cutoff_angle to beam_width (both specified in degrees),
+after which it remains at the maximum value. A projection texture may optionally be supplied.
+
+ */
+
+template <typename Float, typename Spectrum>
+class SpotLight final : public Emitter<Float, Spectrum> {
+public:
+    MTS_IMPORT_BASE(Emitter, m_flags, m_medium, m_world_transform)
+    MTS_IMPORT_TYPES(Scene, Texture)
+
+    SpotLight(const Properties &props) : Base(props) {
+        m_intensity = props.texture<Texture>("intensity", Texture::D65(1.f));
+        m_cutoff_angle = props.float_("cutoff_angle", 20.0f);
+        m_beam_width = props.float_("beam_width", m_cutoff_angle * 3.0f / 4.0f);
+        m_cutoff_angle = deg_to_rad(m_cutoff_angle);
+        m_beam_width = deg_to_rad(m_beam_width);
+        m_inv_transition_width = 1.0f / (m_cutoff_angle - m_beam_width);
+        m_cos_cutoff_angle = cos(m_cutoff_angle);
+        m_cos_beam_width = cos(m_beam_width);
+        Assert(m_cutoff_angle >= m_beam_width);
+        m_uv_factor = tan(m_cutoff_angle);
+        m_texture = props.texture<Texture>("texture", Texture::D65(1.f));
```

Comment:
Let's ensure `intensity` isn't a spatially varying texture and `texture` is when both provided. For this you will need the latest commit on master that adds the `Texture::is_spatially_varying()` method.